packages/headless/src/controllers/commerce/core/url-manager/headless-core-url-manager.test.ts

```diff
@@ -0,0 +1,186 @@
+import {
+  CommerceSearchParameters,
+  restoreSearchParameters,
+} from '../../../../features/commerce/search-parameters/search-parameter-actions';
+import {searchSerializer} from '../../../../features/commerce/search-parameters/search-parameter-serializer';
+import {executeSearch} from '../../../../features/commerce/search/search-actions';
+import {buildMockCommerceState} from '../../../../test/mock-commerce-state';
+import {
+  buildMockCommerceEngine,
+  MockedCommerceEngine,
+} from '../../../../test/mock-engine-v2';
+import {buildSearchParameterManager} from '../../search/parameter-manager/headless-search-parameter-manager';
+import {buildCoreUrlManager, UrlManager} from './headless-core-url-manager';
+
+jest.mock(
+  '../../../../features/commerce/search-parameters/search-parameter-actions'
+);
+jest.mock('../../../../features/commerce/search/search-actions');
+
+describe('core url manager', () => {
+  let engine: MockedCommerceEngine;
+  let manager: UrlManager;
+  let mockedExecuteSearchAction: jest.MockedFunctionDeep<typeof executeSearch>;
+  let mockedRestoreSearchParametersAction: jest.MockedFunctionDeep<
+    typeof restoreSearchParameters
+  >;
+
+  function initEngine(preloadedState = buildMockCommerceState()) {
+    engine = buildMockCommerceEngine(preloadedState);
+  }
+
+  function initUrlManager(fragment = '') {
+    manager = buildCoreUrlManager(engine, {
+      initialState: {
+        fragment,
+      },
+      requestIdSelector: (state) => state.commerceSearch.requestId,
+      parameterManagerBuilder: buildSearchParameterManager,
+      serializer: searchSerializer,
+    });
+  }
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    mockedExecuteSearchAction = jest.mocked(executeSearch);
+    mockedRestoreSearchParametersAction = jest.mocked(restoreSearchParameters);
+    initEngine();
+    initUrlManager();
+  });
+
+  function testLatestRestoreSearchParameters(params: CommerceSearchParameters) {
+    expect(mockedRestoreSearchParametersAction).toHaveBeenLastCalledWith(
+      params
+    );
+  }
+
+  function testExecuteSearch() {
+    expect(mockedExecuteSearchAction).toHaveBeenCalled();
+  }
+
+  describe('initialization', () => {
+    it('dispatches #restoreSearchParameters', () => {
+      expect(mockedRestoreSearchParametersAction).toHaveBeenCalled();
+    });
+
+    it('does not execute a search', () => {
+      expect(mockedExecuteSearchAction).not.toHaveBeenCalled();
+    });
+
+    it('initial #restoreActionCreator should parse the "active" fragment', () => {
+      initUrlManager('q=windmill');
+      testLatestRestoreSearchParameters({
+        q: 'windmill',
+      });
+    });
+  });
+
+  describe('state', () => {
+    it('contains the serialized fragment of the search parameters state', () => {
```

Comment:
Since we have tests for the serializer itself, an alternative would be to test that state.fragment calls the serializer.

Reply:
In line with [Redux's recommendations on testing](https://redux.js.org/usage/writing-tests#selectors), I'd prefer leaving the test as-is. This test ensures the output is as expected, given a certain state.

> Selectors are also generally pure functions, and thus can be tested using the same basic approach as reducers: set up an initial value, call the selector function with those inputs, and **assert that the result matches the expected output**.